fq.go

```diff
@@ -0,0 +1,172 @@
+	return float64(q.clock.Now().UnixNano())
+}
+
+// Enqueue enqueues a packet into the fair queuing scheduler
+func (q *FQScheduler) Enqueue(packet *Packet) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.synctime()
+
+	queue := q.chooseQueue(packet)
+	queue.Enqueue(packet)
+	q.updateTime(packet, queue)
+}
+
+func (q *FQScheduler) getVirtualTime() float64 {
+	return q.vt
+}
+
+func (q *FQScheduler) synctime() {
+	realNow := q.clock.Now()
+	timesincelast := realNow.Sub(q.lastRealTime).Nanoseconds()
+	q.lastRealTime = realNow
+	q.vt += float64(timesincelast) * q.getvirtualtimeratio()
+}
+
+func (q *FQScheduler) getvirtualtimeratio() float64 {
+	NEQ := 0
+	reqs := 0
+	for _, queue := range q.queues {
+		reqs += queue.RequestsExecuting
+		// It might be best to delete this line. If everything is working
+		//  correctly, there will be no waiting packets if reqs < C on current
+		//  line 85; if something is going wrong, it is more accurate to say
+		// that virtual time advanced due to the requests actually executing.
+
+		// reqs += len(queue.Packets)
+		if len(queue.Packets) > 0 || queue.RequestsExecuting > 0 {
+			NEQ++
+		}
+	}
+	// no active flows, virtual time does not advance (also avoid div by 0)
+	if NEQ == 0 {
+		return 0
+	}
+	return min(float64(reqs), float64(C)) / float64(NEQ)
+}
+
+func (q *FQScheduler) updateTime(packet *Packet, queue *Queue) {
+	// When a request arrives to an empty queue with no requests executing
+	// len(queue.Packets) == 1 as enqueue has just happened prior (vs  == 0)
+	if len(queue.Packets) == 1 && queue.RequestsExecuting == 0 {
+		// the queue’s virtual start time is set to getVirtualTime().
+		queue.virstart = q.getVirtualTime()
+	}
+}
+
+// FinishPacketAndDequeue is a convenience method used using the FQScheduler
+// at the concurrency limit
+func (q *FQScheduler) FinishPacketAndDeque(p *Packet) (*Packet, bool) {
+	q.FinishPacket(p)
+	return q.Dequeue()
+}
+
+// FinishPacket is a callback that should be used when a previously dequeud packet
+// has completed it's service.  This callback updates imporatnt state in the
+//  FQScheduler
+func (q *FQScheduler) FinishPacket(p *Packet) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.synctime()
+	S := q.clock.Since(p.startTime).Nanoseconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	q.queues[p.queueidx].virstart -= G - float64(S)
+
+	// request has finished, remove from requests executing
+	q.queues[p.queueidx].RequestsExecuting--
+}
+
+// Dequeue dequeues a packet from the fair queuing scheduler
+func (q *FQScheduler) Dequeue() (*Packet, bool) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	// TODO(aaron-prindle) unsure if this should be here...
+	// q.synctime()
+	queue := q.selectQueue()
+
+	if queue == nil {
+		return nil, false
+	}
+	packet, ok := queue.Dequeue()
+
+	if ok {
+		// When a request is dequeued for service -> q.virstart += G
+		queue.virstart += G
```

Comment:
This should be done whenever a packet is dequeued.  It should be part of `queue.Dequeue()`.